Refresh cpu manager state if inconsistant

peter-wangxu opened the pull request:
The commit tries to solve following 2 cases:
1. The machine has been scaled up, the default cpuset
   should be updated accordingly so that container can use it.
2. The machine has been scaled down, the default cpuset
   should be updated accordingly so that invalid cpuset was
   removed which avoid container startup due to out-of-range cpuset
   for container runtime

In both cases, each container should be added via AddContainer for
potential new cpuset.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #79823

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

peter-wangxu committed:
Refresh cpu manager state if inconsistant

The commit tries to solve following 2 cases:
1. The machine has been scaled up, the default cpuset
   should be updated accordingly so that container can use it.
2. The machine has been scaled down, the default cpuset
   should be updated accordingly so that invalid cpuset was
   removed which avoid container startup due to out-of-range cpuset
   for container runtime

In both cases, each container should be added via AddContainer for
potential new cpuset.

peter-wangxu commented:
/sig node

peter-wangxu commented:
/assign @derekwaynecarr

mattjmcnaughton reviewed (commented):
/unassign
I don't have a ton of context on this part of the code base.

/cc @ConnorDoyle 
Connor, mind taking a look? I think you've worked on this part of the code base in the past :)

matthyx commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -72,7 +72,7 @@ func TestStaticPolicyStart(t *testing.T) {
 			topo:        topoDualSocketHT,
 			stAssignments: state.ContainerCPUAssignments{
-				"0": cpuset.NewCPUSet(0),
+				"conainer0": cpuset.NewCPUSet(0),
```
nit: probably `container0` here

matthyx commented:
/ok-to-test

matthyx commented:
/lint

matthyx commented:
@peter-wangxu please fix your test and linting issue

matthyx commented:
Strange this error... are the tests passing locally?

peter-wangxu commented:
> Strange this error... are the tests passing locally?
@matthyx 
I just rebased and fixed some newly added tests after my initial version. let's see the result

matthyx commented:
/lgtm
